src/cloud/openstack/restapi/custom/api.pm

```diff
@@ -0,0 +1,437 @@
+    my $self  = {};
+    bless $self, $class;
+
+    if (!defined($options{output})) {
+        print "Class Custom: Need to specify 'output' argument.\n";
+        exit 3;
+    }
+    $options{output}->option_exit(short_msg => "Class Custom: Need to specify 'options' argument.")
+        unless $options{options};
+
+    $options{options}->add_options(arguments => {
+        'hostname:s'        => { name => 'hostname', default => '' },
+        'proto:s'           => { name => 'proto', default => 'http' },
+        'insecure'          => { name => 'insecure', default => 0 },
+
+        _services_options(name=>'keystone', port => 5000, endpoint => '/v3/auth/tokens'),
+
+        'username:s'        => { name => 'username', default => '' },
+        'password:s'        => { name => 'password', default => '' },
+        'user-domain:s'     => { name => 'user_domain', default => 'default' },
+        'project-name:s'    => { name => 'project_name', default => 'admin' },
+        'project_domain:s'  => { name => 'project_domain', default => 'default' },
+        'authent-by-env:s'  => { name => 'authent_by_env', default => '0' },
+        'authent-by-file:s' => { name => 'authent_by_file', default => '' },
+
+        'timeout:s'         => { name => 'timeout', default => 10 } } )
+            unless $options{noptions};
+
+    $options{options}->add_help(package => __PACKAGE__, sections => 'REST API OPTIONS', once => 1);
+
+    $self->{output} = $options{output};
+    $self->{http} = centreon::plugins::http->new(%options, default_backend => 'curl');
+
+    $self->{cache_authent} = centreon::plugins::statefile->new(%options);
+
+    return $self;
+}
+
+# Mapping between some OptnStack environment variables and our options
+my %_external_conf_equiv = ( OS_USERNAME => 'username',
+                             OS_PASSWORD => 'password',
+                             OS_PROJECT_DOMAIN => 'project_domain',
+                             OS_USER_DOMAIN => 'user_domain',
+                             OS_PROJECT_NAME => 'project_name',
+                             OS_AUTH_URL => 'keystone_url',
+                           );
+sub apply_external_conf {
+    my ($self, %options) = @_;
+
+    # Some options can be taken from OpenStack environement variables already defined or
+    # taken from a file that has been generated for OpenStack CLI tools
+
+    if ($options{apply_conf_from_env}) {
+        foreach (keys %_external_conf_equiv) {
+            $self->{$_external_conf_equiv{$_}} = $ENV{$_}
+                if exists $ENV{$_};
+        }
+    }
+
+    if ($options{apply_conf_from_file}) {
+        open(my $file, "<".$options{apply_conf_from_file})
+            or $self->{output}->option_exit(short_msg => "Cannot open file '".$options{apply_conf_from_file}."': $!");
+        foreach my $line (<$file>) {
+            next unless $line =~ /^[\s\t]*export[\t\s]+(\w+)=["']?(.*?)["']?$/;
+
+            # Only handle variables defined using 'export variable="value"'
```

Comment:
So the config file only aims at overloading values already available in the environment? Can you add a link to the documentation in the comments?